subsys/bluetooth/host/att.c

```diff
@@ -274,14 +276,94 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf)
 	return err;
 }
 
+static bool att_chan_matches_chan_opt(struct bt_att_chan *chan, enum bt_att_chan_opt chan_opt)
+{
+	__ASSERT_NO_MSG(chan_opt < 0x03);
+
+	if (chan_opt == BT_ATT_CHAN_ANY) {
+		return true;
+	}
+
+	if (atomic_test_bit(chan->flags, ATT_ENHANCED)) {
+		return (chan_opt == BT_ATT_CHAN_ENHANCED);
+	} else {
+		return (chan_opt == BT_ATT_CHAN_UNENHANCED);
+	}
+}
+
+static struct net_buf *get_first_buf_matching_chan(struct k_fifo *fifo,
+						   struct bt_att_chan *chan)
+{
+#if defined(CONFIG_BT_EATT)
+	struct k_fifo skipped;
+	struct net_buf *buf;
+	struct net_buf *ret = NULL;
+
+	k_fifo_init(&skipped);
+
+	while ((buf = net_buf_get(fifo, K_NO_WAIT))) {
+		if (!ret &&
+		    att_chan_matches_chan_opt(
+			    chan, bt_att_tx_meta_data(buf)->chan_opt)) {
+			ret = buf;
+		} else {
+			net_buf_put(&skipped, buf);
+		}
+	}
+
+	__ASSERT_NO_MSG(k_fifo_is_empty(fifo));
+
+	while ((buf = k_fifo_get(&skipped, K_NO_WAIT))) {
```

Comment:
can you measure the processing time of this re-shuffling of the fifo? say for a 10-el queue on an nRF52 chip.

Reply:
I did some basic measurements using `k_uptime_ticks()` on 52840, queuing up a bunch of notifications. With 14 and 15 elements queued, it seems to take about 8 ticks.

Reply:
I guess there can always be a perf PR later if we need it.